deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/api/layers/LayerConstraint.java

```diff
@@ -0,0 +1,11 @@
+package org.deeplearning4j.nn.api.layers;
+
+import org.deeplearning4j.nn.api.Layer;
+import org.nd4j.shade.jackson.annotation.JsonTypeInfo;
+
+@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, include = JsonTypeInfo.As.PROPERTY, property = "@class")
+public interface LayerConstraint {
+
+    void applyConstraint(Layer layer, int iteration, int epoch);
```

Comment:
I see where the call would come in the optimizer, but again it seems dangerous that this could be applied less than once (or more than once, depending on the constraint) per pass. Now that we're tracking epochs, could we have some epoch checking to make sure two successive calls are idempotent and less than one per epoch log a warning?

Reply:
The way I'll be implementing it is that it'll get called once per iteration, after updates have been applied - it won't be possible to apply it more than once per iteration (and, even if it does: all of the implementations will give the same result with multiple sequential appliacions.
The epoch (and, iteration) number is just informational, to allow flexibility for users to have behaviour that depends on the number of epochs passed.